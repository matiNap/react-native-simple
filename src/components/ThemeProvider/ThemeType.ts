interface Palette {
  primary: string;
  secondary: string;
  accent: string;
  text: {
    primary: string;
    secondary: string;
  };
}

interface Typography {
  primaryFont: string | undefined;
  fontSize: {
    header: number;
    small: number;
    normal: number;
    medium: number;
    big: number;
  };
}

export default interface ThemeType {
  palette: Palette;
  paletteDark: Palette;
  typography: Typography;
}
