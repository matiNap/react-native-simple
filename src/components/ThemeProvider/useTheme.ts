import { useContext } from 'react';
import { ThemeContext } from './index';
import { ThemeType } from '../../types';

export default <T>(): T & ThemeType => {
    const theme = useContext<ThemeType & T>(ThemeContext);

    return theme;
};
