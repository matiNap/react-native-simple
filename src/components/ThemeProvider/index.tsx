import React, { ReactNode } from 'react';
import ThemeType from './ThemeType';
import defaultTheme from '../../theme/defaultTheme';

export const ThemeContext = React.createContext<ThemeType>(defaultTheme);

interface Props {
    theme: ThemeType;
    children: ReactNode;
    darkMode: undefined | boolean;
}

const consctructTheme = (theme: ThemeType, key: 'palette' | 'paletteDark') => {
    return {
        ...defaultTheme,
        currentPalette: {
            ...defaultTheme[key],
            ...theme[key],
        },
    };
};

const createCurrentTheme = (
    theme: ThemeType,
    darkMode: boolean | undefined,
) => {
    if (theme && darkMode) {
        return consctructTheme(theme, 'paletteDark');
    } else if (theme && !darkMode) {
        return consctructTheme(theme, 'palette');
    } else if (!theme && darkMode) {
        return consctructTheme(defaultTheme, 'paletteDark');
    } else return consctructTheme(defaultTheme, 'palette');
};

export default ({ theme, children, darkMode }: Props) => {
    const currentTheme = createCurrentTheme(theme, darkMode);

    return (
        <ThemeContext.Provider value={currentTheme}>
            {children}
        </ThemeContext.Provider>
    );
};
