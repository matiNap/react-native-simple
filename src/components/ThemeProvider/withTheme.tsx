import React from 'react';
import useTheme from './useTheme';
import { SimpleTheme } from '../../types';

interface WithThemeProp {
    theme: SimpleTheme;
}

export default <P extends Record<string, unknown>>(
    Parent: React.ComponentType<P>,
) => {
    return (props: P & WithThemeProp) => {
        const theme = useTheme();
        return <Parent {...props} {...{ theme }} />;
    };
};
