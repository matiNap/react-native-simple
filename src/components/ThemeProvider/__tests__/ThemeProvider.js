import 'jsdom-global/register';
import React from 'react';
import { expect, it } from '@jest/globals';
import ThemeProvider from '../index';
import { shallow, mount } from 'enzyme';
import { View } from 'react-native';
import toJson from 'enzyme-to-json';
import useTheme from '../useTheme';
import withTheme from '../withTheme';

describe('Theme provider', () => {
  it('Should override theme with hook', () => {
    const wrapper = shallow(
      <ThemeProvider
        theme={{
          palette: {
            accent: 'red',
          },
        }}
      >
        {() => {
          const theme = useTheme();
          return <View />;
        }}
      </ThemeProvider>,
    );
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should override theme with hoc', () => {
    const Child = withTheme(() => {
      return <View />;
    });
    const wrapper = mount(
      <ThemeProvider
        theme={{
          palette: {
            accent: 'red',
          },
        }}
      >
        <Child />
      </ThemeProvider>,
    );
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
