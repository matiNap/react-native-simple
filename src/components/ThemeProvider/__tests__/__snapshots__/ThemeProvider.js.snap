// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme provider Should override theme with hoc 1`] = `
<_default
  theme={
    Object {
      "palette": Object {
        "accent": "red",
      },
    }
  }
>
  <Component>
    <Component
      theme={
        Object {
          "currentPalette": Object {
            "accent": "red",
            "disabled": Object {
              "primary": "#9BA1A5",
              "secondary": "#c7c8c9",
            },
            "primary": "#ededed",
            "secondary": "#bfbfbf",
            "text": Object {
              "primary": "#333333",
              "secondary": "#f2f2f2",
            },
          },
          "palette": Object {
            "accent": "red",
            "disabled": Object {
              "primary": "#9BA1A5",
              "secondary": "#c7c8c9",
            },
            "primary": "#ededed",
            "secondary": "#bfbfbf",
            "text": Object {
              "primary": "#333333",
              "secondary": "#f2f2f2",
            },
          },
          "paletteDark": Object {
            "accent": "#b626e2",
            "disabled": Object {
              "primary": "#9BA1A5",
              "secondary": "#c7c8c9",
            },
            "primary": "#494949",
            "secondary": "#727272",
            "text": Object {
              "primary": "#f2f2f2",
              "secondary": "#2f2f2f",
            },
          },
          "typography": Object {
            "fontSize": Object {
              "big": 27,
              "medium": 25,
              "normal": 20,
              "small": 15,
            },
            "primaryFont": undefined,
          },
        }
      }
    >
      <View>
        <View />
      </View>
    </Component>
  </Component>
</_default>
`;

exports[`Theme provider Should override theme with hook 1`] = `
<ContextProvider
  value={
    Object {
      "currentPalette": Object {
        "accent": "red",
        "disabled": Object {
          "primary": "#9BA1A5",
          "secondary": "#c7c8c9",
        },
        "primary": "#ededed",
        "secondary": "#bfbfbf",
        "text": Object {
          "primary": "#333333",
          "secondary": "#f2f2f2",
        },
      },
      "palette": Object {
        "accent": "red",
        "disabled": Object {
          "primary": "#9BA1A5",
          "secondary": "#c7c8c9",
        },
        "primary": "#ededed",
        "secondary": "#bfbfbf",
        "text": Object {
          "primary": "#333333",
          "secondary": "#f2f2f2",
        },
      },
      "paletteDark": Object {
        "accent": "#b626e2",
        "disabled": Object {
          "primary": "#9BA1A5",
          "secondary": "#c7c8c9",
        },
        "primary": "#494949",
        "secondary": "#727272",
        "text": Object {
          "primary": "#f2f2f2",
          "secondary": "#2f2f2f",
        },
      },
      "typography": Object {
        "fontSize": Object {
          "big": 27,
          "medium": 25,
          "normal": 20,
          "small": 15,
        },
        "primaryFont": undefined,
      },
    }
  }
>
  <Component />
</ContextProvider>
`;
